keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,117 @@
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Slow starting containers are difficult to address with the current status of health probes: you either kill them before being up, or you have the risk of leaving dead-locked for a very long time.
+
+I am suggesting to add a numerical option `maxInitialFailureCount` to probes to allow a greater number of failures during the initial start of the container before taking action, while keeping `failureThreshold` at a minimum to restart dead-locked containers after an acceptable delay.
+
+## Motivation
+
+The main problem with slow starting containers is that you have to leave them enough time to start before having `livenessProbe` fail `failureThreshold` times (otherwise `kubelet` kills them before they have a chance to be up).
+This is even worse when load on the nodes can increase this long startup time since you have to take into account the worse case scenario (let's call it **T**).
+
+You can handle that today with various strategies:
+
+- set `initialDelaySeconds` > **T**: the `livenessProbe` isn't run for **T** time (so you're blind in the monitoring), you have to wait at least **T** until `kubelet` kills your container after a deadlock - this could only happen at the beginning of the container lifetime
+- set `failureThreshold` * `periodSeconds` > **T**: the `livenessProbe` is run every `periodSeconds` (so you can monitor the probe status), **but you still** have to wait at least **T** until `kubelet` kills your container after a deadlock - this could happen at any time
+
+### Goals
+
+Add an elegant and self-explanatory solution to running slow starting containers in Kubernetes with health probes enabled.
+
+By using `maxInitialFailureCount` * `periodSeconds` > **T** you have the following benefits:
+
+- `livenessProbe` is run every `periodSeconds` - good for monitoring
+- container has enough time to start in the worse case scenario
+- when `livenessProbe` has succeeded once, it has to fail only `failureThreshold` (should be set to 1 or at least low) before `kubelet` kills it
+
+### Non-Goals
+
+Do not debate on whether end-users should or not deploy such containers in Kubernetes.
```

Comment:
I'm not quite sure what this means.  Are you saying that the KEP is for the mechanism, not for policy about whether/when end users should use it?  If so, state it that way.  But further down:

> Documentation must be clearly stating that this feature should only be used in particular cases.

so it sounds like you believe that it should offer a position on when it should be used.  If so, there needs to be a goal to document that position.  If not, just put this simple non-goal here and say nothing more about the subject.

Reply:
You marked this resolved; what was the resolution?

Reply:
goals and non-goals updated (I just haven't pushed)